core/testsuite/src/test/java/com/blazebit/persistence/testsuite/Issue571Test.java

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright 2014 - 2018 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.testsuite;
+
+import com.blazebit.persistence.CTE;
+import com.blazebit.persistence.CriteriaBuilder;
+import com.blazebit.persistence.impl.query.CustomSQLTypedQuery;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoDatanucleus;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoEclipselink;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate42;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate43;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoMySQL;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoOpenJPA;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoOracle;
+import com.blazebit.persistence.testsuite.entity.IdClassEntity;
+import com.blazebit.persistence.testsuite.entity.IdClassEntityId;
+import com.blazebit.persistence.testsuite.tx.TxVoidWork;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import javax.persistence.Entity;
+import javax.persistence.EntityManager;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.IdClass;
+import javax.persistence.ManyToOne;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Objects;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @since 1.3.0
+ * @author Jan-Willem Gmelig Meyling
+ */
+public class Issue571Test extends AbstractCoreTest {
+
+    @Override
+    protected Class<?>[] getEntityClasses() {
+        return new Class<?>[] { Cte.class, MyEntity.class, IdClassEntity.class, IdClassReferencingIdClass.class, EntityWithNestedIdClass.class };
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        transactional(new TxVoidWork() {
+            @Override
+            public void work(EntityManager em) {
+                MyEntity myEntity = new MyEntity();
+                myEntity.setAttribute("attr");
+                em.persist(myEntity);
+            }
+        });
+    }
+
+    @Test
+    @Category({ NoDatanucleus.class, NoEclipselink.class, NoOpenJPA.class, NoMySQL.class, NoOracle.class, NoHibernate42.class, NoHibernate43.class })
```

Comment:
Please add a note about why MySQL, Oracle and Hibernate 4.2 and 4.3 aren't supported.

Reply:
MySQL and Oracle because these issues require CTE's and all tests exclude them.
HH4.2 and 4.3 due to a metamodel bug.

Correct?

Reply:
That's ok, just add a comment noting this.